core/src/main/java/io/grpc/internal/ManagedChannelImpl2.java

```diff
@@ -0,0 +1,873 @@
+   * forceful, the shutdown process is still not instantaneous; {@link #isTerminated()} will likely
+   * return {@code false} immediately after this method returns.
+   */
+  @Override
+  public ManagedChannelImpl2 shutdownNow() {
+    log.log(Level.FINE, "[{0}] shutdownNow() called", getLogId());
+    shutdown();
+    delayedTransport.shutdownNow(SHUTDOWN_NOW_STATUS);
+    channelExecutor.executeLater(new Runnable() {
+        @Override
+        public void run() {
+          if (shutdownNowed) {
+            return;
+          }
+          shutdownNowed = true;
+          maybeShutdownNowSubchannels();
+        }
+      }).drain();
+    return this;
+  }
+
+  @Override
+  public boolean isShutdown() {
+    return shutdown.get();
+  }
+
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+    return terminatedLatch.await(timeout, unit);
+  }
+
+  @Override
+  public boolean isTerminated() {
+    return terminated;
+  }
+
+  /*
+   * Creates a new outgoing call on the channel.
+   */
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> newCall(MethodDescriptor<ReqT, RespT> method,
+      CallOptions callOptions) {
+    return interceptorChannel.newCall(method, callOptions);
+  }
+
+  @Override
+  public String authority() {
+    return interceptorChannel.authority();
+  }
+
+  private class RealChannel extends Channel {
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> newCall(MethodDescriptor<ReqT, RespT> method,
+        CallOptions callOptions) {
+      Executor executor = callOptions.getExecutor();
+      if (executor == null) {
+        executor = ManagedChannelImpl2.this.executor;
+      }
+      StatsTraceContext statsTraceCtx = StatsTraceContext.newClientContext(
+          method.getFullMethodName(), censusFactory, stopwatchSupplier);
+      return new ClientCallImpl<ReqT, RespT>(
+          method,
+          executor,
+          callOptions,
+          statsTraceCtx,
+          transportProvider,
+          scheduledExecutor)
+              .setDecompressorRegistry(decompressorRegistry)
+              .setCompressorRegistry(compressorRegistry);
+    }
+
+    @Override
+    public String authority() {
+      String authority = nameResolver.getServiceAuthority();
+      return checkNotNull(authority, "authority");
+    }
+  }
+
+  /**
+   * Terminate the channel if termination conditions are met.
+   */
+  // Must be run from channelExecutor
+  private void maybeTerminateChannel() {
+    if (terminated) {
+      return;
+    }
+    if (shutdown.get() && subchannels.isEmpty() && oobChannels.isEmpty()) {
+      log.log(Level.FINE, "[{0}] Terminated", getLogId());
+      terminated = true;
+      terminatedLatch.countDown();
+      executorPool.returnObject(executor);
+      // Needed for delivering rejections to new calls after OobChannel is terminated.
+      executor = MoreExecutors.directExecutor();
```

Comment:
I'm uncomfortable with calling `executor` after termination, as generally it is a bug. It also seems insufficient, as there is still the race between `executor` being read during `newCall` and `ClientCallImpl.start()`. If this is temporary until we fix the real bug, then I'm fine with it. Ditto in OobChannel.

The comment here also seems strange. Is the reference to `OobChannel` because it was a copy/paste that wasn't fully updated?